### PR TITLE
Adding LARS to ResNet

### PR DESCRIPTION
**Pull request by pkanwar23**
Adding LARS to ResNet

### CODE REVIEW

File: official/resnet/resnet_run_loop.py [hunk 1]
```diff
@@ -266,14 +266,69 @@ def learning_rate_fn(global_step):
...
+    current_epoch: `Tensor` for current epoch.
+    poly_rate: Polynomial decay rate.
+
+    Returns:
+
+    A scaled `Tensor` for current learning rate.
+    """
+
+    # Learning rate schedule for LARS polynomial schedule
+    if flags.FLAGS.batch_size == 16384:
```
Comment (robieta) [inline official/resnet/resnet_run_loop.py]: Can you make these include the intermediate values? So for instance:
```
if flags.FLAGS.batch_size < 8192:
  plr = 5.0
  w_epochs = 5
elif flags.FLAGS.batch_size < 16384:
  plr = 10.0
  w_epochs = 5
...
```

File: official/resnet/resnet_run_loop.py [hunk 2]
```diff
@@ -266,14 +266,69 @@ def learning_rate_fn(global_step):
...
+    else:
+      plr = 5.0
+      w_epochs = 5
+
+    w_steps = int(w_epochs * batches_per_epoch)
+    wrate = (
+      plr * tf.cast(global_step, tf.float32) / tf.cast(
+          w_steps, tf.float32))
+
+    num_epochs = 90
```
Comment (robieta) [inline official/resnet/resnet_run_loop.py]: You should be able to grab this from FLAGS

File: official/resnet/resnet_run_loop.py [hunk 3]
```diff
@@ -266,14 +266,69 @@ def learning_rate_fn(global_step):
...
+
+  def poly_rate_fn(global_step):
+    """Handles linear scaling rule, gradual warmup, and LR decay.
+
+    The learning rate starts at 0, then it increases linearly per step.  After
+    FLAGS.poly_warmup_epochs, we reach the base learning rate (scaled to account
+    for batch size). The learning rate is then decayed using a polynomial rate
+    decay schedule with power 2.0.
+
+    Args:
```
Comment (dubey) [inline official/resnet/resnet_run_loop.py]: Where are these args?

File: official/resnet/resnet_run_loop.py (hunk 2 above)
Comment (tfboyd) [reply official/resnet/resnet_run_loop.py]: Do we need exact batch-sizes?  Is there a formula you can use?  It seems hard to test if we need to hit/trigger exact batch-sizes, meaning we can only test if we have x GPUs running perf GPU batch of z.  Just asking not saying this is wrong.

File: official/resnet/imagenet_main.py [hunk 4]
```diff
@@ -44,6 +44,19 @@
 
 DATASET_NAME = 'ImageNet'
 
+flags.DEFINE_bool('enable_lars',
```
Comment (robieta) [inline official/resnet/imagenet_main.py]: Can you move these into `define_resnet_flags` in `resnet_run_loop.py`?

File: official/resnet/resnet_run_loop.py (hunk 1 above)
Comment (pkanwar23) [reply official/resnet/resnet_run_loop.py]: done

File: official/resnet/resnet_run_loop.py (hunk 2 above)
Comment (pkanwar23) [reply official/resnet/resnet_run_loop.py]: done. Removed exact batch-sizes too.

File: official/resnet/imagenet_main.py (hunk 4 above)
Comment (pkanwar23) [reply official/resnet/imagenet_main.py]: done

File: official/resnet/resnet_run_loop.py (hunk 3 above)
Comment (pkanwar23) [reply official/resnet/resnet_run_loop.py]: done.

File: official/resnet/resnet_run_loop.py (hunk 2 above)
Comment (robieta) [reply official/resnet/resnet_run_loop.py]: Talked offline, TODO is fine for now.

File: official/resnet/resnet_run_loop.py [hunk 5]
```diff
@@ -266,14 +266,67 @@ def learning_rate_fn(global_step):
...
+    global_step: the current global_step
+
+    Returns:
+    returns the current learning rate
+    """
+
+    # Learning rate schedule for LARS polynomial schedule
```
Comment (robieta) [inline official/resnet/resnet_run_loop.py]: I think this is different from before. Shouldn't it be:
```
if flags.FLAGS.batch_size < 8192:
  plr = 5.0
  w_epochs = 5
elif flags.FLAGS.batch_size < 16384:
  plr = 10.0
  w_epochs = 5
if flags.FLAGS.batch_size < 32768:
  plr = 25.0
  w_epochs = 5
else:
  plr = 32.0
  w_epochs = 14
```

Comment (pkanwar23) [reply official/resnet/resnet_run_loop.py]: facepalm. fixed.

**Comment by robieta** [comment]
Thanks! Enjoy the big shiny green button.

**Pull request merged**